File: hack/openshift/update_bundle.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

const (
	defaultImagePullspec = "registry.redhat.io/bpfman/bpfman-rhel9-operator@sha256:277681e5eecc8ef1c21ec7f0668ba28be660f561c6f46feb0fac3f79b473ab5c"
)

func main() {
	var (
		csvFile       = flag.String("csv-file", "", "Path to the CSV file to update")
		imagePullspec = flag.String("image-pullspec", defaultImagePullspec, "Operator image pullspec")
		outputFile    = flag.String("output", "", "Output file (defaults to input file)")
		dryRun        = flag.Bool("dry-run", false, "Show what would be changed without modifying files")
		help          = flag.Bool("help", false, "Show help message")
	)

	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage: %s [options] <csv-file> [output-file]\n\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "Update operator CSV file with Red Hat branding and metadata.\n\n")
		fmt.Fprintf(os.Stderr, "Options:\n")
		flag.PrintDefaults()
		fmt.Fprintf(os.Stderr, "\nExamples:\n")
		fmt.Fprintf(os.Stderr, "  %s --csv-file input.yaml\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "  %s --dry-run input.yaml\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "  %s input.yaml output.yaml\n", os.Args[0])
	}

	flag.Parse()

	if *help {
		flag.Usage()
		os.Exit(0)
	}

	// Sanitize pullspec by removing leading/trailing whitespace and newlines
	*imagePullspec = strings.TrimSpace(*imagePullspec)

	// Handle positional arguments
	args := flag.Args()
	if *csvFile == "" && len(args) > 0 {
		*csvFile = args[0]
	}
	if *outputFile == "" && len(args) > 1 {
		*outputFile = args[1]
	}

	// Default output to input file if not specified
	if *outputFile == "" {
		*outputFile = *csvFile
	}

	if *csvFile == "" {
		fmt.Fprintf(os.Stderr, "Error: CSV file is required\n\n")
		flag.Usage()
		os.Exit(1)
	}

	if *dryRun {
		fmt.Printf("Dry run mode: would update %s\n", *csvFile)
		if *outputFile != *csvFile {
			fmt.Printf("Output would be written to: %s\n", *outputFile)
		}
		fmt.Printf("Image pullspec: %s\n", *imagePullspec)
	}

	fmt.Printf("Updating CSV file: %s\n", *csvFile)

	content, err := os.ReadFile(*csvFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file %s: %v\n", *csvFile, err)
		os.Exit(1)
	}

	contentStr := string(content)

	replacements := map[string]string{
		"quay.io/bpfman/bpfman-operator:latest": fmt.Sprintf("\"%s\"", *imagePullspec),
		"displayName: Bpfman Operator":          "displayName: eBPF Manager Operator",
		"The bpfman Operator":                   "The eBPF manager Operator",
		"name: The bpfman Community":            "name: Red Hat",
		"url: https://bpfman.io":                "url: https://www.redhat.com",
	}

	if *dryRun {
		fmt.Println("\nString replacements that would be made:")
		for old, new := range replacements {
			if strings.Contains(contentStr, old) {
				fmt.Printf("  %s -> %s\n", old, new)
			}
		}
		fmt.Println("\nWould also add architecture labels and OpenShift feature annotations.")
		os.Exit(0)
	}

	for old, new := range replacements {
		contentStr = strings.ReplaceAll(contentStr, old, new)
	}

	// Parse YAML for structured modifications
	var bpfmanOperatorCSV map[string]interface{}
	err = yaml.Unmarshal([]byte(contentStr), &bpfmanOperatorCSV)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing YAML: %v\n", err)
		os.Exit(1)
	}

	timestamp := time.Now()

	metadata, ok := bpfmanOperatorCSV["metadata"].(map[string]interface{})
	if !ok {
		metadata = make(map[string]interface{})
		bpfmanOperatorCSV["metadata"] = metadata
	}

	labels, ok := metadata["labels"].(map[string]interface{})
	if !ok {
		labels = make(map[string]interface{})
		metadata["labels"] = labels
	}

	annotations, ok := metadata["annotations"].(map[string]interface{})
	if !ok {
		annotations = make(map[string]interface{})
		metadata["annotations"] = annotations
	}

	labels["operatorframework.io/arch.amd64"] = "supported"
	labels["operatorframework.io/arch.arm64"] = "supported"
	labels["operatorframework.io/arch.ppc64le"] = "supported"
	labels["operatorframework.io/arch.s390x"] = "supported"
	labels["operatorframework.io/os.linux"] = "supported"

	annotations["createdAt"] = timestamp.Format("02 Jan 2006, 15:04")
	annotations["features.operators.openshift.io/disconnected"] = "true"
	annotations["features.operators.openshift.io/fips-compliant"] = "true"
	annotations["features.operators.openshift.io/proxy-aware"] = "false"
	annotations["features.operators.openshift.io/tls-profiles"] = "false"
	annotations["features.operators.openshift.io/token-auth-aws"] = "false"
	annotations["features.operators.openshift.io/token-auth-azure"] = "false"
	annotations["features.operators.openshift.io/token-auth-gcp"] = "false"

	output, err := yaml.Marshal(bpfmanOperatorCSV)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshalling YAML: %v\n", err)
		os.Exit(1)
	}

	err = os.WriteFile(*outputFile, output, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error writing file %s: %v\n", *outputFile, err)
		os.Exit(1)
	}

	fmt.Printf("CSV file updated successfully: %s\n", *outputFile)
}

File: hack/openshift/update-bundle.py
#!/usr/bin/env python3

import argparse
import os
import sys
from datetime import datetime
from ruamel.yaml import YAML

yaml = YAML()


def main():
    parser = argparse.ArgumentParser(
        description="Update operator CSV file with Red Hat branding and metadata.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""Examples:
  %(prog)s --csv-file input.yaml --image-pullspec registry.example.com/operator:v1.0.0
  %(prog)s input.yaml output.yaml --image-pullspec registry.example.com/operator:v1.0.0""",
    )

    parser.add_argument("--csv-file", help="Path to the CSV file to update")
    parser.add_argument(
        "--image-pullspec", required=True, help="Operator image pullspec"
    )
    parser.add_argument(
        "--version", help="Version to set in CSV spec.version field"
    )
    parser.add_argument("--output", help="Output file (defaults to input file)")
    parser.add_argument(
        "files",
        nargs="*",
        help="CSV file and optional output file (positional arguments)",
    )

    args = parser.parse_args()

    if args.csv_file is None and len(args.files) > 0:
        args.csv_file = args.files[0]
    if args.output is None and len(args.files) > 1:
        args.output = args.files[1]

    if args.output is None:
        args.output = args.csv_file

    if args.csv_file is None:
        parser.error("CSV file is required")

    args.image_pullspec = args.image_pullspec.strip()

    try:
        with open(args.csv_file, "r") as f:
            content = f.read()
    except FileNotFoundError:
        print(f"Error: File {args.csv_file} not found", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error reading file {args.csv_file}: {e}", file=sys.stderr)
        sys.exit(1)

    # IMPORTANT: Replacement order matters! More specific patterns should come before
    # general ones to avoid conflicts. The script should be run on the original CSV
    # file, not on previously transformed content.
    replacements = [
        ("quay.io/bpfman/bpfman-operator:latest", f'"{args.image_pullspec}"'),
        ("displayName: Bpfman Operator", "displayName: eBPF Manager Operator"),
        ("The bpfman Operator", "The eBPF manager Operator"),
        ("name: The bpfman Community", "name: Red Hat"),
        ("url: https://bpfman.io", "url: https://www.redhat.com"),
        ("https://github.com/bpfman/bpfman", "https://github.com/openshift/bpfman-operator"),
        ("https://github.com/bpfman/bpfman-operator", "https://github.com/openshift/bpfman-operator"),
        ("https://bpfman.netlify.app/", "https://github.com/openshift/bpfman-operator"),
        ("Support bpfman Community", "Support Red Hat"),
        ("bpfman Community", "Red Hat"),
        ("repository: https://github.com/bpfman/", "repository: https://github.com/openshift/bpfman-operator"),
    ]

    for old, new in replacements:
        content = content.replace(old, new)

    try:
        bpfman_operator_csv = yaml.load(content)
    except Exception as e:
        print(f"Error parsing YAML: {e}", file=sys.stderr)
        sys.exit(1)

    timestamp = int(datetime.now().timestamp())
    datetime_time = datetime.fromtimestamp(timestamp)

    if "metadata" not in bpfman_operator_csv:
        bpfman_operator_csv["metadata"] = {}
    if "labels" not in bpfman_operator_csv["metadata"]:
        bpfman_operator_csv["metadata"]["labels"] = {}
    if "annotations" not in bpfman_operator_csv["metadata"]:
        bpfman_operator_csv["metadata"]["annotations"] = {}

    bpfman_operator_csv["metadata"]["labels"][
        "operatorframework.io/arch.amd64"
    ] = "supported"
    bpfman_operator_csv["metadata"]["labels"][
        "operatorframework.io/arch.arm64"
    ] = "supported"
    bpfman_operator_csv["metadata"]["labels"][
        "operatorframework.io/arch.ppc64le"
    ] = "supported"
    bpfman_operator_csv["metadata"]["labels"][
        "operatorframework.io/arch.s390x"
    ] = "supported"
    bpfman_operator_csv["metadata"]["labels"][
        "operatorframework.io/os.linux"
    ] = "supported"

    # Add annotations.
    bpfman_operator_csv["metadata"]["annotations"]["createdAt"] = (
        datetime_time.strftime("%d %b %Y, %H:%M")
    )
    bpfman_operator_csv["metadata"]["annotations"][
        "features.operators.openshift.io/disconnected"
    ] = "true"
    bpfman_operator_csv["metadata"]["annotations"][
        "features.operators.openshift.io/fips-compliant"
    ] = "true"
    bpfman_operator_csv["metadata"]["annotations"][
        "features.operators.openshift.io/proxy-aware"
    ] = "false"
    bpfman_operator_csv["metadata"]["annotations"][
        "features.operators.openshift.io/tls-profiles"
    ] = "false"
    bpfman_operator_csv["metadata"]["annotations"][
        "features.operators.openshift.io/token-auth-aws"
    ] = "false"
    bpfman_operator_csv["metadata"]["annotations"][
        "features.operators.openshift.io/token-auth-azure"
    ] = "false"
    bpfman_operator_csv["metadata"]["annotations"][
        "features.operators.openshift.io/token-auth-gcp"
    ] = "false"

    # Update version if provided
    if args.version:
        version = args.version.strip()
        if "spec" not in bpfman_operator_csv:
            bpfman_operator_csv["spec"] = {}
        bpfman_operator_csv["spec"]["version"] = version
        # Update metadata.name to match version (pattern: bpfman-operator.v<version>)
        bpfman_operator_csv["metadata"]["name"] = f"bpfman-operator.v{version}"

    try:
        if args.output == "-":
            yaml.dump(bpfman_operator_csv, sys.stdout)
        else:
            with open(args.output, "w") as f:
                yaml.dump(bpfman_operator_csv, f)
            print(f"CSV file updated successfully: {args.output}")
    except Exception as e:
        print(f"Error writing file {args.output}: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
